src/flag_gems/ops/weightnorm.py

```diff
@@ -0,0 +1,150 @@
+import logging
+
+import torch
+import triton
+import triton.language as tl
+from ..utils import libentry
+import math
+
+try:
+    from triton.language.extra.cuda.libdevice import rsqrt
+except ImportError:
+    try:
+        from triton.language.math import rsqrt
+    except ImportError:
+        from triton.language.libdevice import rsqrt
+import pytest
+
+
+def cfggen():
+    block_m = [1, 2, 4]
+    block_n = [1024, 2048, 4096]
+    warps = [4, 8, 16]
+    configs = [
+        triton.Config({"BLOCK_ROW_SIZE": m, "BLOCK_COL_SIZE": n}, num_warps=w)
+        for m in block_m
+        for n in block_n
+        for w in warps
+    ]
+    return configs
+
+@libentry()
+@triton.autotune(configs=cfggen(), key=["o_shape0", "o_shape1", "o_shape2"])
+@triton.jit(do_not_specialize=["eps"])
+def weight_norm_kernel(
+    output,
+    v,
+    v_broad,
+    g_broad,
+    o_shape0,
+    o_shape1,
+    o_shape2,
+    v_shape0,
+    v_shape2,
+    vb_shape0,
+    vb_shape1,
+    vb_shape2,
+    gb_shape0,
+    gb_shape1,
+    gb_shape2,
+    v_stride0,
+    v_stride1,
+    v_stride2,
+    vb_stride0,
+    vb_stride1,
+    vb_stride2,
+    gb_stride0,
+    gb_stride1,
+    gb_stride2,
+    eps: tl.constexpr,
+    BLOCK_ROW_SIZE: tl.constexpr,
+    BLOCK_COL_SIZE: tl.constexpr,
+):
+  tid_m = tl.arange(0, BLOCK_ROW_SIZE)[:, None]
+  pid = tl.program_id(axis=0) * BLOCK_ROW_SIZE
+  row_offset = pid + tid_m
+  row_mask = row_offset < o_shape1
+
+  tid_n = tl.arange(0, BLOCK_COL_SIZE)[None, :]
+  v_block = tl.zeros([BLOCK_ROW_SIZE, BLOCK_COL_SIZE], dtype=tl.float32)
+  for base in range(0, v_shape0 * v_shape2, BLOCK_COL_SIZE):
+    col_offset = base + tid_n
+    m_idx = col_offset // v_shape2
+    n_idx = row_offset
+    k_idx = col_offset % v_shape2
+
+    mask = m_idx < v_shape0 and row_mask
+
+    v_offsets = m_idx * v_stride0 + n_idx * v_stride1 + k_idx * v_stride2
+    v_value = tl.load(v + v_offsets, mask = mask)
+    v_block += v_value * v_value
+  v_sum = tl.sum(v_block, axis = 1) + eps
+
+
+  for base in range(0, o_shape0 * o_shape2, BLOCK_COL_SIZE):
+    col_offset = base + tid_n
+    m_idx = col_offset // o_shape2
+    n_idx = row_offset
+    k_idx = col_offset % o_shape2
+
+    mask = m_idx < o_shape0 and row_mask
+
+    v_offsets = (m_idx % vb_shape0) * vb_stride0 + (n_idx % vb_shape1) * vb_stride1 + (k_idx % vb_shape2) * vb_stride2
+    v_value = tl.load(v_broad + v_offsets, mask = mask)
+    v_vec = rsqrt(v_sum[:, None]) * v_value
+
+    g_offset = (m_idx % gb_shape0) * gb_stride0 + (n_idx % gb_shape1) * gb_stride1 + (k_idx % gb_shape2) * gb_stride2
+    g_value = tl.load(g_broad + g_offset, mask = mask)
+    out = v_vec * g_value
+    out_offset = m_idx * o_shape1 * o_shape2 +  n_idx * o_shape2 + k_idx
+    tl.store(output + out_offset, out, mask = mask)
+
+def weight_norm(v, g, dim = 0):
+  logging.debug("GEMS WEIGHTNORM")
+
+  v = v.contiguous()
+  g = g.contiguous()
+  dim_neg = dim - len(v.shape)
```

Comment:
This is not necessary.

Reply:
Having checked the Pytorch codebase, we can safely assume dim is either 0 or the last, clearing the way to simplifying the code below.